### PR TITLE
[travis] Fail early when an invalid composer.json is found

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |

It happens :)

### CODE REVIEW

File: .travis.php [hunk 1]
```diff
@@ -21,6 +21,10 @@
 
     $json = file_get_contents($dir.'/composer.json');
     $package = json_decode($json);
+    if (!$package || '{' !== $json[0]) {
+        echo "Invalid composer.json\n";
```
Comment (xabbuh) [inline .travis.php]: What about also dumping the return value of `json_last_error()` (if available)?

Comment (jeremyFreeAgent) [inline .travis.php]: `json_last_error() != 0` should do the job.

Comment (nicolas-grekas) [inline .travis.php]: No need, we must get an object here, and error reporting is helpless without a custom json parser

Comment (xabbuh) [inline .travis.php]: I don't agree. Why shouldn't we make it easier to spot the error? I mean why should someone has to use another linter just to recognise what is actually wrong with the `composer.json` file?

Comment (nicolas-grekas) [inline .travis.php]: Well, because the native php json parser doesn't tell where the error is?

Comment (xabbuh) [inline .travis.php]: Ah sorry, I meant `json_last_error_msg()` (for which we have to check the PHP version first).

Comment (lyrixx) [inline .travis.php]: You could anyway display the current composer file to help debug.

Comment (nicolas-grekas) [inline .travis.php]: Sure, and it is already displayed :-) see one line up in the file

Comment (lyrixx) [inline .travis.php]: Sorry, I fail to see where the **content** of the json is outputed.

**Comment by nicolas-grekas** [comment]
`composer validate` is now used to display a detailed error message in case an invalid composer.json file is found. E.g.:
```
  "src/Symfony/Component/HttpKernel//composer.json" does not contain valid JSON  
  Parse error on line 37:                                                        
  ...ating": "~2.2",    },    "suggest": {                                       
  ---------------------^                                                         
  Expected: 'STRING' - It appears you have an extra trailing comma               
```

**Comment by jeremyFreeAgent** [comment]
:+1:

**Comment by xabbuh** [comment]
:+1: 

Status: Reviewed

File: .travis.php [hunk 2]
```diff
@@ -19,8 +19,11 @@
     }
     echo "$dir\n";
 
-    $json = file_get_contents($dir.'/composer.json');
-    $package = json_decode($json);
+    $json = ltrim(file_get_contents($dir.'/composer.json'));
```
Comment (xabbuh) [inline .travis.php]: `ltrim()` now isn't really necessary anymore, is it?

Comment (nicolas-grekas) [inline .travis.php]: White spaces are still allowed at the beginning of the file, but a few lines below, I need to get the opening `{` as `$json[0]`. So yes it is still necessary...

Comment (xabbuh) [inline .travis.php]: You had that in the first version of your PR but removed it when switching to using Composer's validate command.

Comment (nicolas-grekas) [inline .travis.php]: the first `{` is still removed here: https://github.com/symfony/symfony/pull/16341/files#diff-d3241953fcadf5f8bc91210b1c064c2aR32

Comment (nicolas-grekas) [inline .travis.php]: the first `{` is still removed here: https://github.com/symfony/symfony/pull/16341/files#diff-d3241953fcadf5f8bc91210b1c064c2aR32

**Comment by TomasVotruba** [comment]
:+1:

**Comment by jakzal** [comment]
:+1:

**Pull request merged**